docs/writing-rules/data-flow/taint-mode.md

```diff
@@ -182,118 +500,124 @@ pattern-propagators:
 
-For example, pattern `$S.add($E)` includes two metavariables `$S` and `$E`. Given `from: $E` and `to: $S`, and with `$E` matching `x` and `$S` matching `s`, when `x` is tainted then `s` becomes tainted (by side-effect) with the same taint as `x`.
+For example:
 
-## Sanitizers
+```yaml
+pattern-propagators:
+  - patterns:
+    - pattern: |
+        if something($FROM):
+          ...
+          $TO()
+          ...
+    from: $FROM
+    to: $TO
+    by-side-effect: false
+```
 
-A sanitizer is specified by a pattern, and **any** subexpression that is matched by this pattern is regarded as sanitized.
+The propagator above specifies that inside an `if` block, where the condition is `something($FROM)`, we want to propagate taint from `$FROM` to any function that is being called without arguments, `$TO()`.
 
-:::note
-Given the sanitizer specification below, and a piece of code such as `sanitize(sink(source))`, the call `sink(source)` is **not** reported.
-The reason is that the pattern `sanitize(...)` matches all of `sanitize(sink(source))`, and that makes Semgrep label every subexpression in that piece of code as being sanitized. In particular, `source` is considered to be sanitized!
+<iframe src="https://semgrep.dev/embed/editor?snippet=4bv6x" border="0" frameBorder="0" width="100%" height="432"></iframe>
 
-```yaml
-pattern-sanitizers:
-- pattern: sanitize(...)
-```
-:::
+Because we are disabling `by-side-effect`, the `sink` occurence that is inside the `if` block is tainted, but this does not affect the `sink` occurrence outside the `if` block.
 
-### Sanitizers by side-effect
+## Findings
 
-Consider the following hypothetical Python code where we sanitize some tainted data by checking whether the data is safe to use, and raising an exception otherwise. It is then guaranteed that after `check_if_safe(x)`, the value of `x` must be a safe one.
+Taint findings are accompanied by a taint trace that explains how the taint flows from source to sink.
 
-```python
-def check_if_safe(x):
-  if not it_is_safe(x):
-    raise Exception("unsafe")
+<!-- <iframe src="https://semgrep.dev/embed/editor?snippet=KxJRL" border="0" frameBorder="0" width="100%" height="432"></iframe> -->
 
-x = tainted
-check_if_safe(x)
-sink(x)
-```
+### Deduplication of findings
+
+Semgrep tracks all the possible ways that taint can reach a sink, but at present it only reports one taint trace among the possible ones. Click "Open in Playground" in the example below, run the example to get one finding, and then ask the Playground to visualize the dataflow of the finding. Even though `sink` can be tainted via `x` or via `y`, the trace will only show you one of these possibiities. If you e.g. replace `x = user_input` with `x = "safe"`, then Semgrep will then report the taint trace via `y`.
```

Comment:
```suggestion
Semgrep tracks all the possible ways that taint can reach a sink, but at present it only reports one taint trace among the possible ones. Click "Open in Playground" in the example below, run the example to get one finding, and then ask the Playground to visualize the dataflow of the finding. Even though `sink` can be tainted via `x` or via `y`, the trace will only show you one of these possibiities. If you replace `x = user_input` with `x = "safe"`, then Semgrep will then report the taint trace via `y`.
```
> Ask the playground to visualize...
Do you mean clicking on Inspect rule in the playground?

Reply:
Nope, clicking on the list of matches (on the right bottom), then click "dataflow" for the match in question.